### PR TITLE
Feature: Recurring Donations Promotion

### PR DESCRIPTION
**Pull request by kjohnson**
## Description

This PR adds the Recurring Donations Promotion when no active gateways support subscriptions (recurring).

## Visuals

image

## Testing Instructions

- Disable the (Next Gen) Test Gateway

### CODE REVIEW

File: packages/form-builder/src/common/getWindowData.ts [hunk 1]
```diff
@@ -6,10 +6,21 @@ declare global {
...
-        };
+        },
+        formBuilderData?: {
+            gateways: Array<any>;
+        },
     }
 }
 
 export default function getWindowData() {
     return window.storageData;
```
Comment (kjohnson) [inline packages/form-builder/src/common/getWindowData.ts:17]: The concept of storage data needs to be refactored as the scope has changed significantly over time.

**Comment by kjohnson** [comment]
Based on @JasonTheAdams [feedback](https://www.loom.com/share/f8f4d7bc5f964e3a9a9425d1a420c148?highlightComment=11411304&t=195) on my weekly [screencast](https://www.loom.com/share/f8f4d7bc5f964e3a9a9425d1a420c148), we might want to consider updating this to take into account supported gateways that are inactive, which will prevent an misleading upsell and could add support for per-form gateways.

> I think the promo should only display if there are no gateways (active or inactive) that support recurring. Otherwise, we risk selling them on something they already have, but just don't have any recurring gateways active. I think that's a niche scenario, but valid nonetheless.
>
> It may be worth anticipating that Per Form Gateways will make this a thing within the Form Builder, as they check which gateways they want this specific form to allow.

File: packages/form-builder/src/common/getWindowData.ts [hunk 2]
```diff
@@ -6,10 +6,21 @@ declare global {
             formDesigns: FormDesign[];
             formPage: FormPageSettings;
             currency: string;
-        };
+        },
+        formBuilderData?: {
+            gateways: Array<any>;
```
Comment (jonwaldstein) [inline packages/form-builder/src/common/getWindowData.ts:11]: let's make a typescript `Gateway` type/interface with id, label, supportsSubscriptions.

Then here we can do `gateways: Gateway[]`

The other areas you are using this should then be aware of the properties of the gateway object.

File: src/FormBuilder/Routes/RegisterFormBuilderPageRoute.php [hunk 3]
```diff
@@ -80,13 +82,25 @@ public function renderPage()
 
         $formBuilderStorage->loadInFooter()->enqueue();
 
+        $builderPaymentGatewayData = array_map(function ($gateway) {
+            return [
+                'id' => $gateway::id(),
+                'label' => give_get_gateway_checkout_label($gateway::id()) ?? $gateway->getPaymentMethodLabel(),
+                'supportsSubscriptions' => $gateway->supportsSubscriptions(),
+            ];
+        }, give(DonationFormRepository::class)->getEnabledPaymentGateways($donationFormId));
```
Comment (jonwaldstein) [inline src/FormBuilder/Routes/RegisterFormBuilderPageRoute.php:91]: Remember we also need to filter these gateways by next gen compatibility or _all_ gateways will be loaded. We should be able to just check if they implement `NextGenPaymentGatewayInterface`

File: packages/form-builder/src/promos/recurring-donations.tsx [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            boxShadow: '0 2px 2px 0 rgba(221, 221, 221, 0.25)',
+            border: 'solid 0.5px var(--givewp-primary-700)'
+        }}
+    >
+        <div style={{}}>
+            {__('Provide donors the option of making flexible recurring donations.', 'give')}
+        </div>
+        <div style={{display: 'flex', justifyContent: 'space-around'}}>
+            <Button href='#' variant={'primary'} style={{backgroundColor:'var(--givewp-primary-500)', padding: '4px 8px', height: 'auto'}}>Upgrade your plan</Button>
+            <Button href='#' variant={'link'} style={{color: 'var(--givewp-gray-100)'}}>Read more</Button>
```
Comment (jonwaldstein) [inline packages/form-builder/src/promos/recurring-donations.tsx:23]: Let's update these links to go to `https://givewp.com/addons/recurring-donations` for now. (open in new tab)

**Review by jonwaldstein — CHANGES_REQUESTED** [review]

File: packages/form-builder/src/promos/recurring-donations.tsx (hunk 4 above)
Comment (kjohnson) [reply packages/form-builder/src/promos/recurring-donations.tsx]: Resolved, see https://github.com/impress-org/givewp-next-gen/pull/140/commits/9279878afb6150856c255a3475d2c965c9a55ffc.

File: src/FormBuilder/Routes/RegisterFormBuilderPageRoute.php (hunk 3 above)
Comment (kjohnson) [reply src/FormBuilder/Routes/RegisterFormBuilderPageRoute.php]: Resolved, see https://github.com/impress-org/givewp-next-gen/pull/140/commits/3ab5ba6f2e9690ee68b62160d72d3ad65582ffcf.

File: packages/form-builder/src/common/getWindowData.ts (hunk 2 above)
Comment (kjohnson) [reply packages/form-builder/src/common/getWindowData.ts]: Resolved, see https://github.com/impress-org/givewp-next-gen/pull/140/commits/936806a7864c0237645d3d4f813d8a76413cda93.

**Comment by kjohnson** [comment]
@jonwaldstein this is ready for review.

**Review by jonwaldstein — APPROVED** [review]
Great job!

**Pull request merged**